Fixed wasm build and improved type safety

pedrovmc opened the pull request:
## :bulb: Motivation and Context
This change was necessary to modernize the web implementation of the PostHog Flutter plugin, replacing the deprecated `dart:js` API with the new `dart:js_interop` API. This fixes WASM compilation issues for Flutter applications using this plugin and improves type safety, reducing the possibility of runtime errors.

The change is crucial for:
1. Ensuring compatibility with newer versions of Flutter and Dart
2. Allowing the plugin to work correctly in WASM environments
3. Improving stability and security with stricter type checking

## :green_heart: How did you test it?
- Tested the plugin compilation for WASM using Flutter 3.19+
- Verified all core plugin functionalities in web environment:
  - Event capturing
  - User identification
  - Feature flags management
  - Reset and session management
- Compared the behavior of the new implementation with the previous implementation to ensure compatibility

## :pencil: Checklist
- [x] I reviewed the submitted code.
- [] I added tests to verify the changes.
- [] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

marandaneto commented:
awesome, thanks @pedrovmc 
i will review it tomorrow first thing.

marandaneto commented on pubspec.yaml:
```diff
@@ -18,6 +18,7 @@ dependencies:
   plugin_platform_interface: ^2.0.2
   # plugin_platform_interface depends on meta anyway
   meta: ^1.3.0
+  web: ^1.1.0
```
this version requires Dart 3.4
so we need to bump this https://github.com/PostHog/posthog-flutter/blob/e44de573ea0aa2068d7c33a839d2c12130b1887c/pubspec.yaml#L10-L11
and then release the SDK as a new major version

marandaneto replied:
min flutter 3.22 https://medium.com/flutter/io24-5e211f708a37 which is not super old.
can we use an older version of `web` so we can keep a better compatibility?
like https://pub.dev/packages/web/versions/0.4.2 uses Dart 3.2, Flutter 3.16

marandaneto replied:
actually i dont see any import from `package:web`, do we actually need it?

marandaneto commented on lib/src/posthog_flutter_web_handler.dart:
```diff
@@ -1,107 +1,180 @@
-// ignore: deprecated_member_use
-import 'dart:js';
+import 'dart:js_interop';
```
do we need to bump the min. version of dart and flutter here?
https://dart.dev/interop/js-interop/package-web#package-web-vs-dart-html
https://github.com/PostHog/posthog-flutter/blob/e44de573ea0aa2068d7c33a839d2c12130b1887c/pubspec.yaml#L10-L11

marandaneto commented on lib/src/posthog_flutter_web_handler.dart:
```diff
@@ -1,107 +1,180 @@
+}
+
+// Acessar o posthog a partir do window
+@JS('window.posthog')
```
if the script fails to set `window.posthog`, this will return null, so then `posthog` should be marked as nullable

marandaneto commented on lib/src/posthog_flutter_web_handler.dart:
```diff
@@ -1,107 +1,180 @@
+}
+
+JSAny mapToJSAny(Map<dynamic, dynamic> map) {
+  return map.jsify()!;
```
fallback to an empty obj instead of forcing and crashing with `!`

marandaneto commented on lib/src/posthog_flutter_web_handler.dart:
```diff
@@ -1,107 +1,180 @@
-
-      return distinctId;
+      final distinctId = posthog.get_distinct_id();
+      return distinctId?.dartify();
```
we didnt do this before but it was not type safety and now it is

marandaneto replied:
also if its empty, we should return null, similar to `getSessionId`

pedrovmc commented:
Hey @marandaneto, thanks for the review! I'll probably go through it by the end of today or tomorrow morning. Thanks again!

marandaneto commented:
> Hey @marandaneto, thanks for the review! I'll probably go through it by the end of today or tomorrow morning. Thanks again!

@pedrovmc nice, just finished the review and tested a few basic things, way to go, thanks and looking forward to your comments/commits.

pedrovmc commented:
## @marandaneto

Thank you for your comments and suggestions! I've made all the requested changes:

1. Removed the `web` package dependency since we're not directly using any imports from it (sorry about that, I was migrating an application that needed it, it's was almost automatic for me).
2. Adjusted minimum SDK requirements to Dart 3.3.0 and Flutter 3.19.0. (https://medium.com/dartlang/history-of-js-interop-in-dart-98b06991158f)
3. Fixed the `mapToJSAny` method to return `map.jsify() ?? JSObject()` instead of using the `!` operator
4. Improved type handling in the `distinctId` method to return `as String?`
5. Added verification to return `null` when `distinctId` is empty, similar to the `getSessionId` method

I've tested all these changes successfully in a WASM environment. The plugin is now more type-safe and fully compatible with the latest Flutter WASM builds.

Please let me know if you need any further adjustments.
